File: java/src/main/java/com/ajax/leetcode/geeks4geeks/QuickSort.java
package com.ajax.leetcode.geeks4geeks;

public class QuickSort {
    public static int[] quicksort(int[] unsorted) {
        if (unsorted == null || unsorted.length < 2) {
            return unsorted;
        }
        int len = unsorted.length;
        int[] sorted = new int[len];
        System.arraycopy(unsorted, 0, sorted, 0, len);
        quicksort(sorted, 0, len - 1);
        return sorted;
    }

    private static void quicksort(int[] array, int low, int high) {
        if (low < high) {
            int pivot = partition(array, low, high);
            quicksort(array, low, pivot - 1);
            quicksort(array, pivot + 1, high);
        }
    }

    private static int partition(int[] array, int low, int high) {
        int index = low - 1;
        for (int i = low; i < high; i++) {
            if (array[i] < array[high]) {
                swap(array, ++index, i);
            }
        }
        swap(array, ++index, high);
        return index;
    }

    private static void swap(int[] array, int x, int y) {
        int temp = array[x];
        array[x] = array[y];
        array[y] = temp;
    }
}

File: go/geeks4geeks/quicksort.go
package main

func QuickSort(unsorted []int) []int {
	var length = len(unsorted)
	if length < 2 {
		return unsorted
	}
	var sorted = make([]int, length)
	copy(sorted, unsorted)
	DoQuickSort(sorted, 0, length)
	return sorted
}

func DoQuickSort(arr []int, low int, high int) {
	if low < high {
		var pi = QuickSortPartition(arr, low, high)
		DoQuickSort(arr, low, pi-1)
		DoQuickSort(arr, pi+1, high)
	}
}

func QuickSortPartition(arr []int, low int, high int) int {
	var pivot = low - 1
	var pivotValue = arr[high-1]
	for i := low; i < high; i++ {
		if arr[i] <= pivotValue {
			pivot += 1
			arr[i], arr[pivot] = arr[pivot], arr[i]
		}
	}
	pivot += 1
	arr[high-1], arr[pivot] = arr[pivot], arr[high-1]
	return pivot
}
